### ISSUE TITLE
Programmatic naming for test suites

### ISSUE THREAD

**Issue by tcfurrer** [gradle/gradle#28938]
### Expected Behavior

Some way to create a set of test suites whose names are given from a programmatically-generated list, instead of being directly coded into the gradle script.

Something like: (notice the [broken] use of "${suiteName}" and iteration over myListOfSuites [created earlier])
```
// See: https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html
def myListOfSuites = ['generated','by','earlier','code']
testing {
    suites {
        test {
            useJUnitJupiter()
        }

        myListOfSuites.each { suiteName ->
            "${suiteName}"Test(JvmTestSuite) {
                targets {
                    all {
                        testTask.configure {
                            description = "Runs all tests tagged with '${suiteName}'"
                            options.includeTags suiteName
                        }
                    }
                }
            }
        }
    }
}
```

### Current Behavior (optional)

Gradle fails with:

```
   > Could not create an instance of type org.gradle.api.plugins.jvm.internal.DefaultJvmTestSuite.
      > Cannot add task 'processTestResources' as a task with that name already exists.
```

### Context

I want to define a set of test suites without hardcoding their names into the groovy code (for example, they might come from parsing an external file).  I have not yet found a way to accomplish this.

**Comment by big-guy** [comment]
This is already possible.  You can try two different ways:
```
        myListOfSuites.each { suiteName ->
            // explicit call to register
            register("${suiteName}Test", JvmTestSuite) {
                targets {
                    all {
                        testTask.configure {
                            description = "Runs all tests tagged with '${suiteName}'"
                            options.includeTags suiteName
                        }
                    }
                }
            }
        }
```
Something like this would work in both Kotlin DSL and Groovy DSL.

or
```
        myListOfSuites.each { suiteName ->
            // implicit call to create
            "${suiteName}Test"(JvmTestSuite) {
                targets {
                    all {
                        testTask.configure {
                            description = "Runs all tests tagged with '${suiteName}'"
                            options.includeTags suiteName
                        }
                    }
                }
            }
        }
```

With the way you have it now, you're creating a test suite named "Test" multiple times.

**Comment by tcfurrer** [comment]
@big-guy Thanks for the help.  However, when using either of the code examples you gave, I'm having a hard time figuring out how to configure the dependencies for the custom test tasks, as described on this page:

https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html#configure_dependencies_of_a_test_suite

I'm successfully using JUnit 5.10.0 (from a local repository because my project must be buildable offline) with the normal test task using that method.  But those dependencies apparently don't automatically apply to the additional test tasks, and using the same syntax for those seems to have no effect.

```
(I'll edit this comment to add the code here soon)
```

When attempting to run the custom test suites with that code, I get the following failure:

<img width="742" alt="image" src="https://github.com/gradle/gradle/assets/39924916/fcf62b6c-1f45-47f6-9224-5a4a3c406b90">

Is this a bug in Gradle, or am I missing something about how to configure dependencies for test suites?

**Comment by big-guy** [comment]
Are you using different coordinates for JUnit?  Those requests are coming from the default dependencies for test suites.

The default test suite (the one named "test") does not automatically add dependencies unless a testing framework is explicitly configured. This was done intentionally to not change behavior.  

All other test suites automatically add dependencies based on which testing framework is being used.  By default, a test suite assumes you're using JUnit Jupiter 5.8.2.  You can change the version with `useJUnitJupiter("5.10.0")`.

**Comment by tcfurrer** [comment]
Thanks again @big-guy for taking a moment to reply here...

We are using JUnit from a local repository because our project needs to be buildable and testable offline (zero internet downloads). Also we need to be in full control of our dependencies, and don't want to use any built-in defaults. Thus I believe we actually must do useJUnitPlatform() instead of useJUnitJupiter().  (Sorry, I should not have shown example code above with useJUnitJupiter... I was confused when I tried that for the simplified example I gave here.  Our actual project has always been doing useJUnitPlatform() .)

The "configure dependencies of a test suite" documentation makes it sounds like suite dependencies are just as configurable as those used for the "test" task.  Any asymmetry in capabilities between the "test" task and other suites seems strange to me.

**_Is it possible to specify local JUnit5 .jar dependencies (in the detailed manner that I showed here) for test suites?_**

(I would be happy to file a new issue with a minimal testcase.  But it seems like a "yes" or "no" answer to that question here might inform the next steps.)

**Comment by big-guy** [comment]
> We are using JUnit from a local repository because our project needs to be buildable and testable offline (zero internet downloads). Also we need to be in full control of our dependencies, and don't want to use any built-in defaults. 

These could still use the usual coordinates though?

> Thus I believe we actually must do useJUnitPlatform() instead of useJUnitJupiter(). 

We're trying to disentangle some past confusion here, so the APIs have similar names, but they mean different things.

The Test task supports the JUnit4, TestNG and JUnitPlatform testing frameworks. JUnit Jupiter is "JUnit5", which is JUnitPlatform + the Jupiter test engine.

The Test task API used to refer to JUnitPlatform as "JUnit5" and has the method `useJUnitPlatform` that's used for JUnit Jupiter and any other JUnitPlatform based testing framework (e.g., Spock 2).

The JvmTestSuite API refers to the testing framework (Spock, JUnit Jupiter, etc), so it uses the API `useJUnitJupiter` and doesn't have a `useJUnitPlatform`. 

Using `useJunitJupiter` with the test suite does a couple of things:
- It configures the Test task to use the appropriate testing framework
- It adds dependencies to the test suite based on which testing framework is being used

This is inherent to how the test suite works.

> The "configure dependencies of a test suite" documentation makes it sounds like suite dependencies are just as configurable as those used for the "test" task. Any asymmetry in capabilities between the "test" task and other suites seems strange to me.

You can add any dependencies you want, but the test suite adds its own based on which testing framework is being used.  This is similar to how the code quality plugins work where Gradle knows which dependencies to add based on just the version.

> Is it possible to specify local JUnit5 .jar dependencies (in the detailed manner that I showed here) for test suites?

No, the coordinates are part of Gradle's assumptions about the testing framework. You would either need to align the coordinates (to match Gradle's) or use something like a ResolutionStrategy/dependency rewriter to switch the coordinates to something else. We've talked about having a way to request/configure a test suite that has no testing framework, but most builds just use JUnit4 or JUnit Jupiter. 

> Maybe this issue is as much related to the fact that we are using the individual .jar files for JUnit5, rather than the monolithic one? 

I'm not sure where you're seeing a monolithic jar? There are several jars that are needed for JUnit5. Maybe you mean that you're explicitly listing all of the dependencies vs listing one and relying on transitive dependencies?

**Comment by big-guy** [comment]
> I realized that I'm OK with using the jupiter monolithic .jar, since we can still use the other .jar files for the platform. The only difference is that the engine will become accessible at build time, which isn't something I wanted to allow, but it's the lesser evil here if useJUnitJupiter() is that is needed to get going here (at least until an enhancement is made to Gradle, assuming the answer to my question is going to be "no" or "not yet").

By "build time" do you mean compile time?  It has to be available for runtime to run the tests...

I think Gradle's doing the right thing:
https://scans.gradle.com/s/dn5jo6dbnza7i/dependencies?dependencies=engine&expandAll

The Jupiter engine is only available at test runtime.

**Comment by tcfurrer** [comment]
Ok so after aligning the dependency coordinates and using useJUnitJupiter("5.10.0"), I got much further on setting up my test suites. I suppose I never should have modified the coordinates in the first place (just a noob mistake with no good justification).

A final thanks to @big-guy for patiently explaining a bunch of things here.  I understand things much better now. A lot of the questions directed at me basically boil down to "oh, you're absolutely right, I muddled that up / used incorrect terms / etc".

My last challenge is to get test suites working together with test fixtures.  I found this helpful discussion which hopefully will help me get unblocked there:
https://discuss.gradle.org/t/how-do-jvm-test-suite-and-java-test-fixtures-relate-to-each-other-and-how-can-they-be-used-cross-project/46491/3

But that issue has nothing to do with this one.  I think I'm done holding discussion on this already-closed issue. Whew!
